File: 3741-minimum-distance-between-three-equal-elements-ii/3741-minimum-distance-between-three-equal-elements-ii.java
import java.util.*;

class Solution {
    public int minimumDistance(int[] nums) {
        Map<Integer, ArrayList<Integer>> pos = new HashMap<>();
        int ans = Integer.MAX_VALUE;

        for (int i = 0; i < nums.length; i++) {
            int x = nums[i];
            pos.putIfAbsent(x, new ArrayList<>());
            ArrayList<Integer> arr = pos.get(x);
            arr.add(i);

            if (arr.size() >= 3) {
                int n = arr.size();
                ans = Math.min(ans, arr.get(n - 1) - arr.get(n - 3));
            }
        }

        return ans == Integer.MAX_VALUE ? -1 : ans * 2;
    }
}

File: leetcode/3741-minimum-distance-between-three-equal-elements-ii/3741-minimum-distance-between-three-equal-elements-ii.ts
function minimumDistance(nums: number[]): number {
    const pos = new Map<number, number[]>();
    let ans = Infinity;

    for (let i = 0; i < nums.length; i++) {
        const x = nums[i];
        if (!pos.has(x)) pos.set(x, []);
        const arr = pos.get(x)!;
        arr.push(i);

        if (arr.length >= 3) {
            ans = Math.min(ans, arr[arr.length - 1] - arr[arr.length - 3]);
        }
    }

    return ans === Infinity ? -1 : ans * 2;
}